rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -49,32 +50,61 @@ def __init__(self) -> None:
         self._use_regex_interpreter = False
         self._default_feature_states = {}
         self.action_texts = []
-        self.entity_tag_id_mapping = {}
+        self.entity_tag_specs = []
 
-    def get_entity_tag_ids(self) -> Dict[Text, int]:
+    def _create_entity_tag_specs(
+        self, bilou_tagging: bool = False
+    ) -> List[EntityTagSpec]:
         """Returns the tag to index mapping for entities.
 
         Returns:
             Tag to index mapping.
         """
         if ENTITIES not in self._default_feature_states:
-            return {}
+            return []
 
-        tag_ids = {
-            tag: idx + 1  # +1 to keep 0 for the NO_ENTITY_TAG
-            for tag, idx in self._default_feature_states[ENTITIES].items()
-        }
-        tag_ids[NO_ENTITY_TAG] = 0
-        return tag_ids
+        if bilou_tagging:
+            tag_id_index_mapping = {
+                f"{prefix}{tag}": idx_1 * len(BILOU_PREFIXES) + idx_2 + 1
+                for tag, idx_1 in self._default_feature_states[ENTITIES].items()
+                for idx_2, prefix in enumerate(BILOU_PREFIXES)
+            }
+        else:
+            tag_id_index_mapping = {
+                tag: idx + 1  # +1 to keep 0 for the NO_ENTITY_TAG
+                for tag, idx in self._default_feature_states[ENTITIES].items()
+            }
+
+        # NO_ENTITY_TAG corresponds to non-entity which should correspond to 0 index
+        # needed for correct prediction for padding
+        tag_id_index_mapping[NO_ENTITY_TAG] = 0
+
+        # TODO
+        #  The entity states used to create the tag-idx-mapping contains the
+        #  entities and the concatenated entity and roles/groups. We do not
+        #  distinguish between entities and roles/groups right now.
+        #  we return a list to anticipate that
```

Comment:
would this be a tuple rather than a list?

Reply:
in line 87 I return a list, do you think I should return a tuple? I think list is better

Reply:
If it's always 2 things, then it should be a tuple right? A list is for a collection of varying length.

Reply:
it's not always 2 things. it can be 1 or 2 or 3 things, and 2nd thing is not always the same.
We have entities, roles and groups, and there might be situations:
- nothing
- only entities
- entities and roles
- entities and groups
- entities and roles and groups

Reply:
ok, so three nullable things then? A dictionary would work? Having a mapping from type of thing to index it occurs if not ideal.

Reply:
maybe. but it is the way it is done in nlu currently. I'd suggest that it should be done in separate PR, since it'll require refactoring a lot of nlu code